consensus/src/quorum_store/utils.rs

```diff
@@ -309,22 +330,33 @@ impl ProofQueue {
         let expiration = proof.expiration();
 
         let batch_sort_key = BatchSortKey::from_info(proof.info());
-        let queue = self.author_to_batches.entry(author).or_default();
-        queue.insert(batch_sort_key.clone(), proof.info().clone());
+        self.author_to_batches
+            .entry(author)
+            .or_default()
+            .insert(batch_sort_key.clone(), proof.info().clone());
         self.expirations.add_item(batch_sort_key, expiration);
-        self.batch_to_proof
-            .insert(batch_key.clone(), Some((proof, Instant::now())));
+        let item = self
+            .items
+            .entry(batch_key)
+            .and_modify(|item| {
+                item.proof = Some(proof.clone());
+                item.proof_insertion_time = Some(Instant::now());
+            })
+            .or_insert_with(|| QueueItem {
+                info: proof.info().clone(),
+                proof: Some(proof),
+                proof_insertion_time: Some(Instant::now()),
+                txn_summaries: None,
+            });
 
-        if let Some((txn_summaries, _expiration, added)) = self.batch_summaries.get_mut(&batch_key)
-        {
-            if !*added {
-                for txn_summary in txn_summaries {
-                    *self
-                        .txn_summary_num_occurrences
-                        .entry(*txn_summary)
-                        .or_insert(0) += 1;
-                }
-                *added = true;
+        // If we are here, then proof is added for the first time. Otherwise, we will
+        // return early. We only count when proof is added for the first time.
+        if let Some(ref txn_summaries) = item.txn_summaries {
```

Comment:
If the proof is added multiple times, this logic is also triggered multiple times right?
Earlier, I used a flag for each batch_key to indicate whether the `txn_summaries` has been added to `txn_summary_num_occurrences` already. May be we should add a similar boolean value in `QueueItem`?